### PR TITLE
Biome Adding stuff

### PR DESCRIPTION
**Pull request by jonahnm**
### Changes made in this pull request
No direct changes, mostly adding new things (this likely still needs more testing before merge, but I don't have enough time to do extensive testing)

### CODE REVIEW

File: SMLHelper.sln [hunk 1]
```diff
@@ -39,20 +29,6 @@ Global
...
-		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.SN.EXP|Any CPU.ActiveCfg = SN.EXP|Any CPU
-		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.SN.EXP|Any CPU.Build.0 = SN.EXP|Any CPU
-		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.SN.STABLE|Any CPU.ActiveCfg = SN.STABLE|Any CPU
-		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.SN.STABLE|Any CPU.Build.0 = SN.STABLE|Any CPU
-		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.BZ.EXP|Any CPU.ActiveCfg = BZ.STABLE|Any CPU
-		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.BZ.STABLE|Any CPU.ActiveCfg = BZ.STABLE|Any CPU
-		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.BZ.STABLE|Any CPU.Build.0 = BZ.STABLE|Any CPU
-		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.SN.EXP|Any CPU.ActiveCfg = SN.STABLE|Any CPU
-		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.SN.STABLE|Any CPU.ActiveCfg = SN.STABLE|Any CPU
-		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.SN.STABLE|Any CPU.Build.0 = SN.STABLE|Any CPU
```
Comment (PrimeSonic) [inline SMLHelper.sln:55]: Revert all changes to this file

File: SMLHelper/BiomeThings/BiomeClass.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                    result.Add(biome.batchIds[e]);
+                }
+            }
+            return result;
+        }
+        internal static Texture2D finalbiomemap = SMLHelper.V2.Utility.ImageUtils.LoadTextureFromFile(Path.Combine(SNUtils.unmanagedDataDir,"Build18","biomeMap.png"));
+        internal static int finalbiomemapheight => finalbiomemap.height;
+        internal static int finalbiomemapwidth => finalbiomemap.width;
+        internal static bool ignoreHeader = false;
+        internal static List<Color> usedrandomcolors = new List<Color>();
```
Comment (PrimeSonic) [inline SMLHelper/BiomeThings/BiomeClass.cs:105]: Are these members used anywhere?

If not, please remove.
If they are, then use [PascalCase](https://khalilstemmler.com/blogs/camel-case-snake-case-pascal-case/) for consistency with naming conventions.

File: SMLHelper/BiomeThings/BiomeClass.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        public abstract Texture2D EditedBiomeMap(out int height, out int width);
+        /// <summary>
+        /// The gameobject with the collider for registering when biome is entered
+        /// </summary>
+        /// <returns>A gameobject with collider applied</returns>
+        public abstract GameObject GetCollider();
+        /// <summary>
+        /// The Terrain for each batch of your biome, in format (BatchId,Terrain)
+        /// </summary>
+        public abstract Dictionary<Int3, GameObject> batchTerrains { get; }
```
Comment (PrimeSonic) [inline SMLHelper/BiomeThings/BiomeClass.cs:81]: All `public`, `protected, and `internal` members should be in _PascalCase_.

Use _camelCase_ only for `private` members

File: SMLHelper/BiomeThings/FloatingOrigin.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        if (UpdateAllScenes)
+        {
+            for (int z = 0; z < SceneManager.sceneCount; z++)
+            {
+                foreach (GameObject g in SceneManager.GetSceneAt(z).GetRootGameObjects())
+                {
+                    if(g.name == "MainCamera (UI)" || g.name == "MainCamera")
+                    {
+                        continue;
+                    }else if(g.TryGetComponent<uGUI_BuilderMenu>(out var buildermenu))
```
Comment (PrimeSonic) [inline SMLHelper/BiomeThings/FloatingOrigin.cs:72]: Please run a the automatic code formatting on all updated and new files.

This will minimize unintended whitespace changes going forward.

File: SMLHelper/Initializer.cs [hunk 5]
```diff
@@ -6,7 +6,7 @@
     using Patchers;
     using Patchers.EnumPatching;
     using QModManager.API.ModLoading;
-
+    using System.IO;
```
Comment (PrimeSonic) [inline SMLHelper/Initializer.cs:9]: Remove unused `using` statements.

File: SMLHelper/SMLHelper.csproj [hunk 6]
```diff
@@ -72,12 +72,10 @@
...
     </Reference>
     <Reference Include="Assembly-CSharp-firstpass_publicized">
-      <HintPath>$(Dependencies)\Assembly-CSharp-firstpass_publicized.dll</HintPath>
-      <Private>False</Private>
+      <HintPath>D:\SteamLibrary\steamapps\common\Subnautica\Subnautica_Data\Managed\publicized_assemblies\Assembly-CSharp-firstpass_publicized.dll</HintPath>
     </Reference>
     <Reference Include="Assembly-CSharp_publicized">
-      <HintPath>$(Dependencies)\Assembly-CSharp_publicized.dll</HintPath>
-      <Private>False</Private>
+      <HintPath>D:\SteamLibrary\steamapps\common\Subnautica\Subnautica_Data\Managed\publicized_assemblies\Assembly-CSharp_publicized.dll</HintPath>
```
Comment (PrimeSonic) [inline SMLHelper/SMLHelper.csproj:78]: Revert these changes

File: SMLHelper/SMLHelper.csproj [hunk 7]
```diff
@@ -88,7 +86,24 @@
...
+    <Reference Include="System.Numerics" />
+    <Reference Include="System.Numerics.Vectors, Version=4.1.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>..\packages\System.Numerics.Vectors.4.5.0\lib\net46\System.Numerics.Vectors.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Runtime.CompilerServices.Unsafe, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>..\packages\System.Runtime.CompilerServices.Unsafe.5.0.0\lib\net45\System.Runtime.CompilerServices.Unsafe.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Text.Encoding.CodePages, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>..\packages\System.Text.Encoding.CodePages.5.0.0\lib\net461\System.Text.Encoding.CodePages.dll</HintPath>
+    </Reference>
```
Comment (PrimeSonic) [inline SMLHelper/SMLHelper.csproj:106]: Are these packages needed?

**Review by PrimeSonic — CHANGES_REQUESTED** [review]
Left comments and questions

File: SMLHelper/SMLHelper.csproj (hunk 7 above)
Comment (jonahnm) [reply SMLHelper/SMLHelper.csproj:106]: These are leftovers from my image experimenting (although unsafe is required)

Comment (jonahnm) [reply SMLHelper/SMLHelper.csproj:106]: It's part of a function signature

File: SMLHelper/BiomeThings/BiomeClass.cs (hunk 2 above)
Comment (jonahnm) [reply SMLHelper/BiomeThings/BiomeClass.cs:105]: These are also leftovers from image experimenting things.

File: SMLHelper/Patchers/PlayerPatcher.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+namespace SMLHelper.V2.Patchers
+{
+    internal class PlayerPatcher
+    {
+            [HarmonyPatch(typeof(Player), nameof(Player.Awake))]
+            [PatchUtils.Postfix]
+            internal static void Player_Awake_Postfix(Player __instance)
+            {
+                __instance.gameObject.EnsureComponent<FloatingOrigin>().ReferenceObject = __instance.transform;
+            }
```
Comment (Metious) [inline SMLHelper/Patchers/PlayerPatcher.cs:18]: Only add the component when the "floating origin" option is checked.

File: SMLHelper/Patchers/PlayerPatcher.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            [PatchUtils.Prefix]
+            internal static bool Player_SetPosition_Prefix(ref Vector3 wsPos)
+            {
+                wsPos -= FloatingOrigin.CurrentOffset;
+                return true;
+            }
+            [HarmonyPatch(typeof(Player), nameof(Player.UpdateBiomeRichPresence))]
+            [PatchUtils.Postfix]
+            public static void Player_UBRP_Postfix(string biomeStr)
+            {
```
Comment (Metious) [inline SMLHelper/Patchers/PlayerPatcher.cs:29]: This patch should be removed since every biome should already have an atmosphere volume and that will take care of the RP things.

File: SMLHelper/Patchers/PrefabDatabasePatcher.cs [hunk 10]
```diff
@@ -14,33 +14,33 @@
 
     internal static class PrefabDatabasePatcher
     {
-        private static class PostPatches
-        {
-            [PatchUtils.Postfix]
-            [HarmonyPatch(typeof(PrefabDatabase), nameof(PrefabDatabase.LoadPrefabDatabase))]
-            internal static void LoadPrefabDatabase_Postfix()
+            private static class PostPatches
```
Comment (Metious) [inline SMLHelper/Patchers/PrefabDatabasePatcher.cs:17]: Revert all of the changes to this file

File: SMLHelper/Patchers/WaterBiomeManagerPatcher.cs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            internal static bool WaterBiomeManager_Start_Prefix(WaterBiomeManager __instance)
+            {
+                for (var i = 0; i < BiomeThings.Variables.Biomes.Count; i++)
+                {
+                    var biome = BiomeThings.Variables.Biomes[i];
+                    var settings = new WaterBiomeManager.BiomeSettings()
+                    {
+                        name = biome.BiomeName,
+                        settings = biome.WaterScapeSettings,
+                        skyPrefab = new UnityEngine.GameObject("Useless Junk.")
```
Comment (Metious) [inline SMLHelper/Patchers/WaterBiomeManagerPatcher.cs:23]: Please dont implicitly add "useless junk" skies. Consumers must have full control over setting up their skies so there should be a `SetupBiomeSettings` method in the `Biome` class, then we add them to the `WaterBiomeManager.biomeSettings`, `WaterBiomeManager.biomeLookup` and `WaterBiomeManager.skies` here.

File: SMLHelper/Patchers/WaterBiomeManagerPatcher.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+                    };
+                    __instance.biomeSettings.Add(settings);
+                }
+                return true;
+            }
+            [HarmonyPatch(typeof(WaterBiomeManager), nameof(WaterBiomeManager.Start))]
```
Comment (Metious) [inline SMLHelper/Patchers/WaterBiomeManagerPatcher.cs:29]: Remove this patch - `AtmosphereVolume` already covers it.

File: SMLHelper/Patchers/WaterBiomeManagerPatcher.cs [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+                .RemoveInstruction()
+                .SetOpcodeAndAdvance(OpCodes.Ldloc_0)
+                .SetAndAdvance(OpCodes.Call, AccessTools.Method(typeof(WaterBiomeManagerPatcher), nameof(WaterBiomeManagerPatcher.GetSkyReplace)))
+                .InstructionEnumeration();
+            }
+
+
+
+            
+                internal static mset.Sky GetSkyReplace(MarmoSkies __instance, Int32 index)
```
Comment (Metious) [inline SMLHelper/Patchers/WaterBiomeManagerPatcher.cs:56]: Remove this patch - `AtmosphereVolume` already covers it.

File: SMLHelper/Patchers/AtmosphereVolumePatcher.cs [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+                for(var i = 0; i < list.Count;i++)
+                {
+                    if (list[i].Calls(typeof(Component).GetMethod(nameof(Component.GetComponent),Type.EmptyTypes).MakeGenericMethod(typeof(Collider))))
+                    {
+                        list[i].operand = typeof(AtmosphereVolumePatcher).GetMethod(nameof(AtmosphereVolumePatcher.GetComponentReplace),System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Static);
+                    }
+                }
+                return list.AsEnumerable();
+            }
+            internal static unsafe Collider GetComponentReplace(Component __instance)
```
Comment (Metious) [inline SMLHelper/Patchers/AtmosphereVolumePatcher.cs:27]: What does this patch do? Seems unnecessary.

File: SMLHelper/BiomeThings/FloatingOrigin.cs [hunk 15]
```diff
@@ -0,0 +1,148 @@
+﻿// Based on the Unity Wiki FloatingOrigin script by Peter Stirling
+// URL: http://wiki.unity3d.com/index.php/Floating_Origin
+
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+internal class FloatingOrigin : MonoBehaviour
```
Comment (Metious) [inline SMLHelper/BiomeThings/FloatingOrigin.cs:7]: This looks very expensive. Please make a toggle option for `FloatingOrigin` and default it to `false`.

File: SMLHelper/BiomeThings/FloatingOrigin.cs [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+
+    private void MoveRootTransforms(Vector3 offset)
+    {
+        if (UpdateAllScenes)
+        {
+            for (int z = 0; z < SceneManager.sceneCount; z++)
+            {
+                foreach (GameObject g in SceneManager.GetSceneAt(z).GetRootGameObjects())
+                {
+                    if (g.name == "MainCamera (UI)" || g.name == "MainCamera")
```
Comment (Metious) [inline SMLHelper/BiomeThings/FloatingOrigin.cs:69]: This code is never reached as these cameras are parented to the player.

**Review by Metious — CHANGES_REQUESTED** [review]
Dont forget to add some comments to your code.

File: SMLHelper/Patchers/PlayerPatcher.cs (hunk 9 above)
Comment (jonahnm) [reply SMLHelper/Patchers/PlayerPatcher.cs:29]: I also want people to be able to have custom rich presences.

File: SMLHelper/BiomeThings/FloatingOrigin.cs (hunk 16 above)
Comment (jonahnm) [reply SMLHelper/BiomeThings/FloatingOrigin.cs:69]: MainCamera (UI) isn't. I've already checked that.

File: SMLHelper/Patchers/WaterBiomeManagerPatcher.cs (hunk 11 above)
Comment (jonahnm) [reply SMLHelper/Patchers/WaterBiomeManagerPatcher.cs:23]: The useless junk sky is almost immediately destroyed and replaced with a provided sky by the consumer. I only add the useless junk sky as to pass a check.

File: SMLHelper/Patchers/EntitySlotGeneratorPatcher.cs [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            {
+                var list = new List<CodeInstruction>(list_);
+                for(var i = 0; i < list.Count;i++)
+                {
+                    if (list[i].Calls(typeof(EntitySlotGenerator.GeneratorRule).GetMethod(nameof(EntitySlotGenerator.GeneratorRule.Match),System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)))
+                    {
+                        list[i].operand = typeof(EntitySlotGeneratorPatcher).GetMethod(nameof(EntitySlotGeneratorPatcher.MatchReplace), System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Static);
+                    }
+                }
+                return list.AsEnumerable();
```
Comment (PrimeSonic) [inline SMLHelper/Patchers/EntitySlotGeneratorPatcher.cs:31]: `AsEnumerable` call is redundant.  
`List<CodeInstruction>` already implements `IEnumerable<CodeInstruction>`

File: SMLHelper/Patchers/EnumPatching/EntityCellPatcher.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+        [HarmonyPatch(typeof(EntityCell),nameof(EntityCell.SerializeWaiterDataAsync))]
+        internal static class EntityCell_SWDA_Patch
+        {
+            [HarmonyTranspiler]
+            internal static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> list_)
+            {
+                var list = new List<CodeInstruction>(list_);
+                for(var i = 0;i < list.Count;i++)
+                {
+                    if (i == 51)
```
Comment (PrimeSonic) [inline SMLHelper/Patchers/EnumPatching/EntityCellPatcher.cs:20]: Is there any better way to identify which instruction to replace beyond counting?  
If this method ever updates, this patch will break.

File: SMLHelper/SMLHelper.csproj [hunk 19]
```diff
@@ -71,14 +71,6 @@
...
       <Private>False</Private>
     </Reference>
-    <Reference Include="Assembly-CSharp-firstpass_publicized">
-      <HintPath>$(Dependencies)\Assembly-CSharp-firstpass_publicized.dll</HintPath>
-      <Private>False</Private>
-    </Reference>
-    <Reference Include="Assembly-CSharp_publicized">
-      <HintPath>$(Dependencies)\Assembly-CSharp_publicized.dll</HintPath>
-      <Private>False</Private>
-    </Reference>
```
Comment (PrimeSonic) [inline SMLHelper/SMLHelper.csproj:81]: Revert this removal

File: SMLHelper/Patchers/PlayerPatcher.cs (hunk 9 above)
Comment (Metious) [reply SMLHelper/Patchers/PlayerPatcher.cs:29]: That can be done by adding a language line. the key should be `PresenceExploring_biome_{BIOME_NAME}` and the game will automatically set its value as RP when the player is there.

File: SMLHelper/BiomeThings/FloatingOrigin.cs (hunk 15 above)
Comment (jonahnm) [reply SMLHelper/BiomeThings/FloatingOrigin.cs:7]: Alright.

File: SMLHelper/Patchers/PlayerPatcher.cs (hunk 9 above)
Comment (jonahnm) [reply SMLHelper/Patchers/PlayerPatcher.cs:29]: Ah, alright.

File: SMLHelper/Patchers/BuilderPatcher.cs [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+using HarmonyLib;
+namespace SMLHelper.V2.Patchers
+{
+    internal class BuilderPatcher
+    {
+            [HarmonyPatch(typeof(Builder), nameof(Builder.UpdateAllowed))]
+            [PatchUtils.Postfix]
+            internal static void Builder_UpdateAllowed_Postfix(ref bool __result)
+            {
+                if(BiomeThings.Variables.Biomes.Exists(biome => biome.BiomeName == Player.main.GetBiomeString() && Player.main.IsInSub() is false))
```
[the remark this answers is not in the document]
Comment (jonahnm) [reply SMLHelper/Patchers/BuilderPatcher.cs:15]: No I have not, I'll try that

File: SMLHelper/Patchers/AtmosphereVolumePatcher.cs (hunk 14 above)
Comment (jonahnm) [reply SMLHelper/Patchers/AtmosphereVolumePatcher.cs:27]: The patch makes it so that the Collider for the AtmosphereVolume and the collider for the terrain isn't confused by the game using GetComponent<Collider>()

File: SMLHelper/BiomeThings/FloatingOrigin.cs (hunk 15 above)
Comment (jonahnm) [reply SMLHelper/BiomeThings/FloatingOrigin.cs:7]: I tested this and it's well noticeable even at 3000,0,3000.

File: SMLHelper/Patchers/EntitySlotGeneratorPatcher.cs (hunk 17 above)
Comment (jonahnm) [reply SMLHelper/Patchers/EntitySlotGeneratorPatcher.cs:31]: Does calling it have a performance impact? If so I will remove the call.

File: SMLHelper/BiomeThings/FloatingOrigin.cs (hunk 15 above)
Comment (Metious) [reply SMLHelper/BiomeThings/FloatingOrigin.cs:7]: It's about as noticeable as when you're at 2000 axis of any direction. Which is only noticeable when you're paying real attention to what's flickering (beginning of (noticeable) precision loss is at around 2000, then at 5000 everything starts going crazy and unplayable). 
Again, this is expensive and not any pc can handle it smoothly so a toggle would be great.

Comment (jonahnm) [reply SMLHelper/BiomeThings/FloatingOrigin.cs:7]: One problem with that, for some reason without origin shifting, entities stop spawning, and I'm not sure why this is.

Comment (jonahnm) [reply SMLHelper/BiomeThings/FloatingOrigin.cs:7]: I'll add the toggle nonetheless though.

Comment (jonahnm) [reply SMLHelper/BiomeThings/FloatingOrigin.cs:7]: Oh apparently I already added it and I forgot.

File: SMLHelper/Patchers/PrefabDatabasePatcher.cs (hunk 10 above)
Comment (jonahnm) [reply SMLHelper/Patchers/PrefabDatabasePatcher.cs:17]: I'm slightly confused, because on my end, it doesn't appear I changed anything.

File: SMLHelper.sln [hunk 21]
```diff
@@ -5,16 +5,6 @@ VisualStudioVersion = 17.1.32210.238
...
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Example mod", "Example mod\Example mod.csproj", "{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}"
-	ProjectSection(ProjectDependencies) = postProject
-		{418502DD-372D-4EF9-8021-B262552DFEDE} = {418502DD-372D-4EF9-8021-B262552DFEDE}
-	EndProjectSection
-EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "SMLHelper.Tests", "SMLHelper.Tests\SMLHelper.Tests.csproj", "{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}"
-	ProjectSection(ProjectDependencies) = postProject
-		{418502DD-372D-4EF9-8021-B262552DFEDE} = {418502DD-372D-4EF9-8021-B262552DFEDE}
-	EndProjectSection
-EndProject
```
Comment (toebeann) [inline SMLHelper.sln:17]:
```suggestion
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Example mod", "Example mod\Example mod.csproj", "{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}"
	ProjectSection(ProjectDependencies) = postProject
		{418502DD-372D-4EF9-8021-B262552DFEDE} = {418502DD-372D-4EF9-8021-B262552DFEDE}
	EndProjectSection
EndProject
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "SMLHelper.Tests", "SMLHelper.Tests\SMLHelper.Tests.csproj", "{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}"
	ProjectSection(ProjectDependencies) = postProject
		{418502DD-372D-4EF9-8021-B262552DFEDE} = {418502DD-372D-4EF9-8021-B262552DFEDE}
	EndProjectSection
EndProject
```

File: SMLHelper.sln (hunk 1 above)
Comment (toebeann) [inline SMLHelper.sln:55]:
```suggestion
		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.BZ.EXP|Any CPU.ActiveCfg = BZ.EXP|Any CPU
		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.BZ.EXP|Any CPU.Build.0 = BZ.EXP|Any CPU
		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.BZ.STABLE|Any CPU.ActiveCfg = BZ.STABLE|Any CPU
		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.BZ.STABLE|Any CPU.Build.0 = BZ.STABLE|Any CPU
		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.SN.EXP|Any CPU.ActiveCfg = SN.EXP|Any CPU
		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.SN.EXP|Any CPU.Build.0 = SN.EXP|Any CPU
		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.SN.STABLE|Any CPU.ActiveCfg = SN.STABLE|Any CPU
		{C8FB0981-77D2-47C7-BBEF-A3A9EBACACBF}.SN.STABLE|Any CPU.Build.0 = SN.STABLE|Any CPU
		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.BZ.EXP|Any CPU.ActiveCfg = BZ.STABLE|Any CPU
		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.BZ.STABLE|Any CPU.ActiveCfg = BZ.STABLE|Any CPU
		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.BZ.STABLE|Any CPU.Build.0 = BZ.STABLE|Any CPU
		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.SN.EXP|Any CPU.ActiveCfg = SN.STABLE|Any CPU
		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.SN.STABLE|Any CPU.ActiveCfg = SN.STABLE|Any CPU
		{EBC4AAF3-A13B-427E-BF99-3AE23EDEB4B8}.SN.STABLE|Any CPU.Build.0 = SN.STABLE|Any CPU
```

File: SMLHelper/Patchers/EntitySlotGeneratorPatcher.cs (hunk 17 above)
Comment (toebeann) [reply SMLHelper/Patchers/EntitySlotGeneratorPatcher.cs:31]:
```suggestion
                return list;
```

File: SMLHelper/Patchers/PrefabDatabasePatcher.cs (hunk 10 above)
Comment (toebeann) [reply SMLHelper/Patchers/PrefabDatabasePatcher.cs:17]: You seem to have inadvertantly added a single indentation to every line inside PrefabDatabasePatcher. Likely your IDE is showing them as the same, as the only change is whitespace, but they are not.

File: SMLHelper/SMLHelper.csproj [hunk 22]
```diff
@@ -72,12 +72,10 @@
       <Private>False</Private>
     </Reference>
     <Reference Include="Assembly-CSharp-firstpass_publicized">
-      <HintPath>$(Dependencies)\Assembly-CSharp-firstpass_publicized.dll</HintPath>
-      <Private>False</Private>
+      <HintPath>D:\SteamLibrary\steamapps\common\Subnautica\Subnautica_Data\Managed\publicized_assemblies\Assembly-CSharp-firstpass_publicized.dll</HintPath>
```
Comment (toebeann) [inline SMLHelper/SMLHelper.csproj:76]:
```suggestion
      <HintPath>$(Dependencies)\Assembly-CSharp-firstpass_publicized.dll</HintPath>
      <Private>False</Private>
```

You should move these DLLs into the dependencies folder of the solution, rather than hard-coding their path into the repository to a file location that may be different for every other contributor.

File: SMLHelper/SMLHelper.csproj (hunk 6 above)
Comment (toebeann) [inline SMLHelper/SMLHelper.csproj:80]:
```suggestion
      <HintPath>$(Dependencies)\Assembly-CSharp_publicized.dll</HintPath>
      <Private>False</Private>
```

File: SMLHelper/BiomeThings/FloatingOrigin.cs [hunk 23]
```diff
@@ -0,0 +1,147 @@
+﻿// Based on the Unity Wiki FloatingOrigin script by Peter Stirling
+// URL: http://wiki.unity3d.com/index.php/Floating_Origin
+
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+internal class FloatingOrigin : MonoBehaviour
```
Comment (toebeann) [inline SMLHelper/BiomeThings/FloatingOrigin.cs:7]: No namespace?

File: SMLHelper/BiomeThings/BiomeClass.cs [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using UnityEngine;
+using Logger = QModManager.Utility.Logger;
+using UWE;
+using WorldStreaming;
+using System.Runtime.CompilerServices;
+namespace SMLHelper.V2.BiomeThings
```
Comment (toebeann) [inline SMLHelper/BiomeThings/BiomeClass.cs:13]: Not the biggest fan of this namespace. Isn't consistent with SMLHelper's naming conventions. Perhaps this should be moved into the `Assets` namespace?

**Review by toebeann — CHANGES_REQUESTED** [review]
This is a preliminary review, having gone in depth on what the code does, merely looking over the formatting, syntax, naming conventions, etc.

**Comment by toebeann** [comment]
You should also please target the `dev` branch, not `master`. We only merge into `master` from `dev`.

File: SMLHelper/BiomeThings/BiomeClass.cs (hunk 24 above)
Comment (PrimeSonic) [reply SMLHelper/BiomeThings/BiomeClass.cs:13]: Yeah, this could be moved into `SMLHelper.V2.Assets.Biomes`

**Comment by jonahnm** [comment]
> You should also please target the `dev` branch, not `master`. We only merge into `master` from `dev`.

Done.

File: SMLHelper.sln (hunk 21 above)
Comment (toebeann) [reply SMLHelper.sln:17]: These projects still need to be added back to the solution, as far as I can tell.

File: SMLHelper.sln (hunk 1 above)
Comment (toebeann) [reply SMLHelper.sln:55]: These solution configs need to be added back as well.

File: SMLHelper/BiomeThings/FloatingOrigin.cs (hunk 23 above)
Comment (jonahnm) [reply SMLHelper/BiomeThings/FloatingOrigin.cs:7]: I'm going to remove this system, as it's very buggy.

File: SMLHelper/Patchers/WaterBiomeManagerPatcher.cs (hunk 12 above)
Comment (jonahnm) [reply SMLHelper/Patchers/WaterBiomeManagerPatcher.cs:31]: so is everything in this file redundant?

**Comment by jonahnm** [comment]
Changed to draft as I need to implement a replacement for floating origin.

**Comment by jonahnm** [comment]
Why has this been closed? I still plan to finish this.

**Comment by Metious** [comment]
It was done automatically because the dev branch got removed and re-added due to merge to main. Pull the new changes onto your fork and then feel free to re-open the PR.
